File: Leetcode/0977-SquaresOfASortedArray/cpp/solution.cpp
#include <cmath>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n);
        int i = 0, j = n-1;
        for (int k = n-1; k >= 0; k--) {
            if (abs(nums[i]) < abs(nums[j])) {
                res[k] = nums[j] * nums[j];
                j--;
            } else {
                res[k] = nums[i] * nums[i];
                i++;
            }
        }
        return res;
    }
};

File: Leetcode/0977-SquaresOfASortedArray/java/Solution.java
class Solution {
    public int[] sortedSquares(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];         // result
        int i = n-1;                    // writing position of res
        int left = 0, right = n-1;
        while (left <= right) {
            if (Math.abs(nums[left]) > Math.abs(nums[right])) {
                res[i] = nums[left]*nums[left];
                left++;
            } else {
                res[i] = nums[right]*nums[right];
                right--;
            }
            i--;
        }
        return res;
    }
}